### PR TITLE
Documentation: Scala tips

### PR DESCRIPTION
**Pull request by REASY**
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/466)

### CODE REVIEW

File: docs/developers.rst [hunk 1]
```diff
@@ -329,3 +329,108 @@ This code marks the test with `com.beam.tags.Periodic` tag. You can also specify
 You can find details about scheduling a continuous integration build under DevOps section `Configure Periodic Jobs`_.
 
 .. _Configure Periodic Jobs: http://beam.readthedocs.io/en/latest/devops.html#configure-periodic-jobs
+
+Scala tips
+^^^^^^^^^^
+Scala Collection
+~~~~~~~~~~~~~~~~
+
+Use ``mutable`` buffer instead of ``immutable var``:
```
Comment (JustinPihony) [inline docs/developers.rst]: I was going to comment on this earlier, but waffled on it a bit due to our allocation issues - should we point out that var in general should be avoided, preferring val?

File: docs/developers.rst [hunk 2]
```diff
@@ -329,3 +329,108 @@ This code marks the test with `com.beam.tags.Periodic` tag. You can also specify
...
+   var buffer = scala.collection.immutable.Vector.empty[Int]
+   buffer = buffer :+ 1
+   buffer = buffer :+ 2
+
+   // After
+   val buffer = scala.collection.mutable.ArrayBuffer.empty[Int]
+   buffer += 1
+   buffer += 2
+
+Don’t create temporary collections, use ``view``:
```
Comment (JustinPihony) [inline docs/developers.rst]: I don't know if this needs to be something written. At minimum it should be described a bit more - maybe point to https://www.scala-lang.org/blog/2017/11/28/view-based-collections.html

File: docs/developers.rst [hunk 3]
```diff
@@ -329,3 +329,108 @@ This code marks the test with `com.beam.tags.Periodic` tag. You can also specify
...
+
+   // Before
+   seq.filter(predicate).headOption
+
+   // After
+   seq.collectFirst { case num if predicate(num) => num }
+
+Don’t emulate ``collect``:
+**************************
+
```
Comment (JustinPihony) [inline docs/developers.rst]: IMHO this can be combined with `collectFirst`. They are the same concept.

File: docs/developers.rst [hunk 4]
```diff
@@ -329,3 +329,108 @@ This code marks the test with `com.beam.tags.Periodic` tag. You can also specify
...
+   val s: Seq[String] = Seq("C#", "C++", "C", "Scala", "Haskel")
+   val predicate: String => Boolean = (s: String)  => { s.size > 3 }
+
+   // Before
+   s.filter(predicate).map { s => s.head }
+
+   // After
+   s.collect { case curr if predicate(curr) => curr.head }
+
+Prefer not to use ``_1, _2,...`` for ``Tuple`` to improve readability:
```
Comment (JustinPihony) [inline docs/developers.rst]: Should this be boiled down to prefer pattern matching? And should there be an example of `val` extraction? 
i.e `val (key, value) = tuple2`

File: docs/developers.rst (hunk 1 above)
Comment (REASY) [reply docs/developers.rst]: For me it's like `common sense` when you code in Scala, that's why I did not mention it here. Should I?

File: docs/developers.rst (hunk 4 above)
Comment (REASY) [reply docs/developers.rst]: Added an example with destructive binding for tuple

File: docs/developers.rst (hunk 1 above)
Comment (JustinPihony) [reply docs/developers.rst]: If you come from Java then it is not common sense. I can go either way, though.

**Comment by REASY** [comment]
Last thing: should I mention other's companies guidelines, for example:
- [databricks](https://github.com/databricks/scala-style-guide)
- [Effective Scala by twitter](http://twitter.github.io/effectivescala/)

**Pull request merged**